Convert dynamic targets 1 to static (#19)
USGS-R/ds-pipelines-targets-1-course: course-instructions.md

<details>
<summary><h2>Pipelines Overview</h2></summary>

**Welcome to the first installment of "introduction to data pipelines"** at USGS!! :sparkles:

Data analyses are often complex. Data pipelines are ways of managing that complexity. Our data pipelines have two foundational pieces:

* Good organization of code scripts helps you quickly find the file you need, whether you or a teammate created it.

* Dependency managers such as `remake`, `scipiper`, `snakemake`, `drake`, and `targets` formalize the relationships among the datasets and functions to ensure reproducibility while also minimizing the amount of unnecessary runtime as you're creating or modifying parts of the pipeline.

</details>

<hr> 

<details>

> <summary><h2>Getting Started With USGS Data Science Pipelines</h2></summary>

Author: Currently the doc is linked at line 212 [here](https://github.com/padilla410/ds-pipelines-targets-1-course-static/blob/main/course-instructions.md?plain=1#L212). I will also link at the top of the doc

<h3>:keyboard: Installing tools</h3>

Before getting started, you'll need some R packages. You will need to install `targets` and its companion package, `tarchetypes`. 

```r
install.packages(c("targets", "tarchetypes"))
```

While you are at it, please install a few other packages that you'll use along the way:

```r
install.packages(c("tidyverse", "sbtools", "whisker", "dataRetrieval"))
```

> When all packages are installed, check the version of the `targets` package with `packageVersion("targets")`, and verify that the number you just output is greater than or equal to `0.3.1`. 

Author: Opened in issue #20

If not, try re-installing:

```r
install.packages(c("targets", "tarchetypes"))
library(targets)
```

If you can't get it to work, try contacting your designated course instructor!

</details>

<hr>

<details>
<summary><h2>Create a Branch in Your Code Repository</h2></summary>

You'll be revising files in this repository shortly. To follow our team's standard git workflow, you should first clone this training repository to your local machine so that you can make file changes and commits there. 

> Open a git bash shell (Windows) or a terminal window (Mac) and change (`cd`) into the directory you work in for projects in R (for me, this is `~/Documents/R`). There, clone the repository and set your working directory to the new project folder that was created:

Author: I am going to push up a commit to do this, but should we be so explicit here? For example, different folks might have differing opinions on where to put stuff. For example, I organize by project so my directory would probably be something like `~/Documents/projects/training` for this one.

I'm on the fence about whether to do this:
```
cd Documents/R
git clone git@github.com:<username>/ds-pipelines-targets-1-course.git
cd ds-pipelines-targets-1-course
```

Or this:
```
cd <your preferred working directory>
git clone git@github.com:<username>/ds-pipelines-targets-1-course.git
cd ds-pipelines-targets-1-course
``

> Open a git bash shell (Windows) or a terminal window (Mac) and change (`cd`) into the directory you work in for projects in R (for me, this is `~/Documents/R`). There, clone the repository and set your working directory to the new project folder that was created:

Author: I went with the latter

While you are at it, this is a good time to invite a few collaborators to your repository, which will make it easier to assign them as reviewers in the future. In the :gear: Settings widget at the top of your forked repo, select "Manage access". Go ahead and invite your course labmate(s). It should look something like this: 
![add some friends](https://user-images.githubusercontent.com/2349007/81471981-c0094900-91ba-11ea-93b0-0ffd31ec4ea9.png)

You can move on from this issue when you've successfully pushed your branch to remote and added some collaborators.

</details>

<hr>

<details>
<summary><h2>Organize Your Project Files</h2></summary>

You should organize your code into functions, targets, and conceptual "phases" of work. 

Often we create temporary code or are sent scripts that look like `my_work_R/my_happy_script.R` in this repository. Take a minute to look through that file now. 

> This code has some major issues, including that it uses a directory that is specific to a user, it plots to a non-project file location, and the structure of the code makes it hard to figure out what is happening. This simple example is a starting point for understanding the investments we make to move towards code that is more reproducible, more shareable, and understandable. Additionally, we want to structure our code and our projects in a way where we can build on top of them as the projects progress. 

Author: I'm going to open this up as an issue - to be completed after this initial PR (#21)

<hr> 

> We're going to combine several tasks into this assignment. You'll be asked to make some modifications to your training repository and also to create a [pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests) that captures these changes. 

Author: I bumped it up into the intro and then add a lil transition text

### Background on phases 

> We use team conventions for how our pipelines are organized, which make it easier to hop in and out of collaborative projects and to rapidly understand what is going on where. 

Author: Adding to #21

We refer to major elements of a pipeline as "phases", and name phases according to their purpose, such as `1_fetch` or `2_process`. These phases are used to separate files and data based on the _intent_ of the code we are writing, and make it tractable to figure out where you'd need to edit code if you were coming in fresh to the project.

> For medium to large pipelines projects, you'll see these workflow phases explicitly named by a number often followed by a verb (separated by an underscore). We use these phases to create different folders :file_folder: for data and code, and also to specify how we orchestrate the running of code (more on that later).

Author: Yes, I like having numbered phases, but i typically do `01` and `02`, etc. I think it is also helpful to have numbers when we start using different names for phases (which is sometimes necessary)

So, if we have a `1_fetch` phase, code in the fetch folder :file_folder: would be used to do things like get data from web services, google drive, an FTP, or to scrape a website. `2_process` (or `2_munge`) might contain code that transforms the "fetched" data into more usable formats. 

We recommend having `src` and `out` folders within each phase folder that contain code _for_ this phase (`src`) and data (or other files) produced _by_ this phase (`out`). When seeing some of our existing pipelines in action, you will also see other folders :file_folder: named `in`, `log`, and `tmp` to represent manually added files, logged/diagnostic output, and temporary data files, respectively.

> 

Author: #22

### :keyboard: Activity: Restructure your code repository to follow our team's conventions for folders and files

Create a two phase directory structure for "fetch" and "process" concepts, and include `src` and `out` folders. Move the example script (`my_happy_script.R`) from the `my_work_R` folder into one of the `src` folders (at this time, it doesn't matter which one you choose) and delete any existing folders that aren't part of the intended structure.

> When you are done, open a `pull request` with the changes.

Author: yes, I think so, right now that is documented in issue #18 for now

<hr> 

Great, your PR is open! Let's do some more work before merging it. Now that your files are organized into phases, next you will add a commit to your pull request that makes changes to the code itself.

### Background on functions 

In addition to phases, it is important to decompose high-level concepts (or existing scripts) into thoughtful functions and _targets_ that form the building blocks of data processing pipelines. In addition to being the name of the pipeline package, a target is a noun we use to describe a tangible output of function, often a file or an R object, that we can use as an end-product (like a summary map), or as an input into another function. 

> We strive to create _functions_ that declare their clear purpose (combining good function naming with thoughtful arguments/inputs is helpful for this) and are designed for re-use when appropriate. When writing pipelines functions, look for areas of re-usable operations or places where simple dividing lines can be drawn between different parts of data access, processing, modeling/analysis, and visualization. We use the high-level "phases" to divide the *major* concepts, but the way we scope functions is an additional subdivide. It is a best practice to have a function do a single thing, so instead of creating two plots and a table, it might be better to use one function to generate a table, which is then used as input to another function to create a plot. There are exceptions to this pattern (a 1:1 function-to-target pairing) that we'll get into later. 

Author: Added to #21

### Setting up a `targets` pipeline

In addition to phases (which we covered earlier), it is important to decompose high-level concepts (or existing scripts) into thoughtful functions and "targets" that form the building blocks of data processing pipelines. A target is a noun we use to describe a tangible output of a function, which is often a file or an R object. Targets can be used as an end-product (like a summary map) or as input _into_ another function to create _another_ target. 

To set up a `targets` pipeline, you will need to create the base makefile named `_targets.R` that will declare and orchestrate the rest of the pipeline connections.

---
A simple version of `_targets.R` might look something like this:

```r
library(targets)
source("code.R")
tar_option_set(packages = c("tidyverse", "sbtools", "whisker"))

list(
  tar_target(
    model_RMSEs_csv,
    download_data(out_filepath = "model_RMSEs.csv"),
    format = "file"
  ), 
  tar_target(
    eval_data,
    process_data(in_filepath = model_RMSEs_csv),
  ),
  tar_target(
    figure_1_png,
    make_plot(out_filepath = "figure_1.png", data = eval_data), 
    format = "file"
  )
)
```

> This file defines the relationships between different "targets" (see how the target `model_RMSEs_csv` is an input to the command that creates the target `eval_data`?), tells us where to find any functions that are used to build targets (see the `source` call that points you to `code.R`), and declares the package dependencies needed to build the different targets (see the `target_option_set()` command that passes in a vector of packages). 

Reviewer: Break this sentence up

Author: done!

We'll briefly explain some of the functions and conventions used here. For more extensive explanations, visit the [`targets` documentation](https://books.ropensci.org/targets/walkthrough.html). 

>   - As you would with normal R scripts, put any `source` commands for loading R files and `library` commands for loading packages at the top of the file. The packages loaded here should be only those needed to build the targets _plan_; packages needed to build specific _targets_ can be loaded later.

Reviewer: "As you would with normal R scripts" is making assumptions about the trainee and what is "normal"

Author: I do think this falls into the category of "normal" - esp. with a "lazy loading" language like R, however I see your point and have removed that assumption.

>   - As you would with normal R scripts, put any `source` commands for loading R files and `library` commands for loading packages at the top of the file. The packages loaded here should be only those needed to build the targets _plan_; packages needed to build specific _targets_ can be loaded later.

Reviewer: ```suggestion
  - As you would with normal R scripts, put any `source` commands for loading R files and `library` commands for loading packages at the top of the file. The packages loaded in `tar_option_set()` should be only those needed to build the targets _plan_; packages needed to build specific _targets_ are loaded later.
```

  - Declare each target by using the function `tar_target()` and passing in a target name (`name` arg) and the expression to run to build the target (`command` arg).

>   - There are two types of targets - objects and files. If your target is a file, you need to add `format = "file"` to your `tar_target()` call and the command needs to return the filename of the new file.

Reviewer: ```suggestion
  - There are two types of targets - objects and files. If your target is a file, you need to add `format = "file"` to your `tar_target()` call and the command needs to return the filename of the new file. `figure_1_png` above is a file target.
```

```
git checkout main
git pull
git branch -d structure
git checkout -b makefile
git push -u origin makefile 
```
---
Next, create the file with the contents we've given you by entering the following from your repo directory in terminal/command line:
```
cat > _targets.R
library(targets)
source("code.R")
tar_option_set(packages = c("tidyverse", "sbtools", "whisker"))

list(
  # Get the data from ScienceBase
  tar_target(
    model_RMSEs_csv,
    download_data(out_filepath = "model_RMSEs.csv"),
    format = "file"
  ), 
  # Prepare the data for plotting
  tar_target(
    eval_data,
    process_data(in_filepath = model_RMSEs_csv),
  ),
  # Create a plot
  tar_target(
    figure_1_png,
    make_plot(out_filepath = "figure_1.png", data = eval_data), 
    format = "file"
  ),
  # Save the processed data
  tar_target(
    model_summary_results_csv,
    write_csv(eval_data, file = "model_summary_results.csv"), 
    format = "file"
  ),
  # Save the model diagnostics
  tar_target(
    model_diagnostic_text_txt,
    generate_model_diagnostics(out_filepath = "model_diagnostic_text.txt", data = eval_data), 
    format = "file"
  )
)
```    

> then use `Ctrl+D` to exit the file creation mode and return to the prompt. 

Reviewer: This is fine, but curious as to why this is done in terminal

Author: This is a hold over from the dynamic training and was probably part of the original scipiper training. I think it is probably a verrry gentle introduction to using `vim` to do stuff.

Finally, create a `pull request` that includes this new file (the file should be called `_targets.R`). In the following section, we'll provide some helpful suggestions to improve your pipeline.

</details>

<hr>

<details>
<summary><h2>Modifying Your Pipeline by Adding to Your Pull Request</h2></summary>

> * Great work, but it would be great to download the `model_RMSEs.csv` to `1_fetch/out/model_RMSEs.csv` instead. Can you make a change to make sure that happens when you build this target?

Reviewer: ```suggestion
* Great work, but now you've realized you want to download the `model_RMSEs.csv` to `1_fetch/out/model_RMSEs.csv` instead. Can you make a change to make sure that happens when you build this target?
```

* Like with the downloaded data, let's move this plot output file into the appropriate phase folder (so, something like `3_visualize/out/figure_1.png`). You can also rename the figure if you'd like to use something more descriptive. 

<hr> 

So, what does this do for you? Well, if you had this `_targets.R` file in your current working directory (which you do), and you had defined the `download_data()`, `process_data()`, `plot_data()`, and `generate_diagnostics()` functions in `code.R` (which you haven't), it would look like this when you ran `tar_make()` from the `targets` package:

![targets run image](https://user-images.githubusercontent.com/13220910/128781588-64911332-6464-4cb3-b77d-eaa7063d6bc1.png)

---

> Now - adding on to this PR by pushing up additional commits - add and modify code, folders, and this makefile so you can build _your_ `3_visualize/out/figure_1.png` with `targets::tar_make()` (BTW, if you haven't run into this syntax, {package_name}::{function_name} allows you to run a function without `library({package_name})`). Build on your existing folders, functions, and code from the previous sections by continuing with the phases `1_fetch`, `2_process`, and `3_visualize`. Comment in the pull request with a screenshot of the build message, like we have in the message above :point_up:. Assign your course contact to review your PR, and they may ask for a few changes before merging. 

Reviewer: First sentence unclear

Author: Added a commit to clarify language here

> There is a file in your repo called `.gitignore`. Add a line with `_targets` so that git doesn't track the contents of the `targets` directory. In addition, make sure that the contents of `1_fetch/out/*` will be ignored by git because you don't want to store changes to your downloaded .csv files (you may have already added a more generic `*/out/*` to your `.gitignore` that handles all `out/` folders in this repo). 

Reviewer: Wonder if this should be like "Whoops! Did you commit your local `_targets` folder? This folder should be ignored..."

Reviewer: Often I am seeing people committing this first, and then not being able to uncommit it when they realize it should be ignored

Author: Yes, It think most people make a mistake here. `gitignore` first gets mentioned at [line 125](https://github.com/padilla410/ds-pipelines-targets-1-course-static/blob/main/course-instructions.md?plain=1#L125). I think it ultimately happens because [line 298](https://github.com/padilla410/ds-pipelines-targets-1-course-static/blob/b4133804c68e11abcf2cbd02482fff0cc6452f96/course-instructions.md?plain=1#L298) says "add and modify code, folders, and this makefile so you can build _your_ `3_visualize/out/figure_1.png` with `targets::tar_make()`".

I think the time to address the `_targets` subfolder is right after the first `tar_make` command.

Author: modified
